File: python/153_find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if nums[0] < nums[-1] or len(nums) == 1:
            return nums[0]

        left, right = 0, len(nums) - 1

        while left < right:
            mid = (left + right) // 2

            if nums[mid] > nums[mid + 1]:
                return nums[mid + 1]
            if nums[mid - 1] > nums[mid]:
                return nums[mid]

            if nums[left] > nums[mid]:
                right = mid - 1
            else:
                left = mid + 1

File: go/153_find-minimum-in-rotated-sorted-array.go
// Binary Search with Recursion
func findMin(nums []int) int {
	left, right := 0, len(nums)-1

	if nums[left] < nums[right] || len(nums) == 1 {
		return nums[0]
	}

	mid := int((left + right) / 2)

	if nums[right] < nums[mid] {
		return findMin(nums[(mid + 1):])
	} else {
		return findMin(nums[:(mid + 1)])
	}
}

/*---------------------------------------------------------------------------------------*/

// Simple Binary Search
func findMin(nums []int) int {
	left, right := 0, len(nums)-1

	for left < right {
		mid := int((left + right) / 2)

		if nums[right] < nums[mid] {
			left = mid + 1
		} else {
			right = mid
		}
	}
	return nums[left]
}